// @ig3/node-jieba-js/src/jieba.mjs
import { load as loadDict, loadSync as loadDictSync } from './loader.mjs';
import * as path from 'path';
import * as fs from 'node:fs';
import * as os from 'node:os';

const reEng = /[a-zA-Z0-9]/;

function cut (sentence, opts = {}, cb) {
  if (typeof opts === 'function') {
    cb = opts;
    opts = {};
  }

  if (this._initPending) throw new Error('initializing');

  const ret = this._cut(sentence, opts);
  if (cb) cb(ret);
  return ret;
}

function useDict (dict) {
  const self = this;

  if (typeof dict === 'string') {
    this.options.dictionaryFiles.push(dict);
    this._buildPfdictFromSourceSync();
  } else if (Array.isArray(dict)) {
    if (!this.options.dictionaryEntries) this.options.dictionaryEntries = [];
    for (let i = 0; i < dict.length; i++) {
      this.options.dictionaryEntries.push(dict[i]);
    }
    this._buildPfdictFromSourceSync();
  } else if (typeof dict === 'function') {
    const ret = dict.call(this, this.dictionary, this);

    if (ret instanceof Promise) {
      return ret.then(dict => {
        if (dict) return self.useDict(dict);
      });
    } else if (ret) {
      self.useDict(ret);
    }
  } else {
    throw new Error('Unsupported dictionary type: ' + typeof dict);
  }

  return this;
}

function getDict () {
  return this.dictionary;
}

function getDictFiles () {
  return this.options.dictionaryFiles;
}

async function init () {
  // If init() is already running, wait for it to finish
  while (this._initPending) await this._initPending;
  // Run _setup() from possibly updated dictionary data
  this._initPending = this._setup();
  await this._initPending;
  delete this._initPending;
}

function initSync () {
  this._setupSync();
}

function _cut (sentence, opts = {}) {
  sentence = sentence.replace(/\r\n?|\n/g, '\n');

  const yieldValues = [];

  const reHan = /([\u4E00-\u9FA5a-zA-Z0-9+#&._]+)/;
  const reSkip = /(\r\n|\s)/;

  const cutter = opts.cutAll
    ? this.__cutAll
    : opts.HMM
      ? this.__cut_DAG
      : this.__cutDagNoHmm;

  const blocks = sentence.split(reHan);

  for (let b = 0; b < blocks.length; b++) {
    const blk = blocks[b];
    if (blk.length === 0) continue;

    // Odd indexes are separators
    if (b % 2) {
      const segregation = cutter(this, blk);
      for (let i = 0; i < segregation.length; i++) {
        yieldValues.push(segregation[i]);
      }
    } else {
      const tmp = blk.split(reSkip);
      for (let i = 0; i < tmp.length; i++) {
        const x = tmp[i];
        if (x.length === 0) continue;
        // Odd indexes are separators
        if (i % 2) yieldValues.push(x);
        else if (opts.cutAll) yieldValues.push(...x);
        else yieldValues.push(x);
      }
    }
  }
  return yieldValues;
}

// At each position, output all matching multi-character words from the
// dictionary. Output the single character only if it is not otherwise
// part of a word. Coalesce runs of consecutive 'English' characters
// that are not part of a Chinese word.
function __cutAll (inst, sentence) {
  const len = sentence.length;
  const pfdict = inst._cache_.pfdict;
  const words = [];
  let last = -1;
  let eng = '';
  for (let i = 0; i < len; i++) {
    const tmp = [];
    for (let j = i + 1; j < len; j++) {
      const word = sentence.slice(i, j + 1);
      const score = pfdict[word];
      if (score === undefined) break;
      if (score) {
        tmp.push(word);
        if (last < j) last = j;
      }
    }
    if (tmp.length) {
      if (eng.length) {
        words.push(eng);
        eng = '';
      }
      words.push(...tmp);
    } else if (last < i) {
      if (reEng.test(sentence[i])) {
        eng += sentence[i];
      } else {
        if (eng.length) {
          words.push(eng);
          eng = '';
        }
        words.push(sentence[i]);
      }
    }
  }
  if (eng.length) words.push(eng);
  return words;
}

// Returns an array of words as strings: a segregation of the given sentence
// into words. It uses essentially the same algorithm as the Python Jieba
// module cut function, except that it does not require the DAG data
// structure. Instead, it works directly from the prefix dictionary.
//
// The default frequency, for any character not in the prefix dictionary
// is the minimum frequency that is in the dictionary.
//
function __cutDagNoHmm (inst, sentence) {
  const len = sentence.length;
  const pfdict = inst._cache_.pfdict;
  const minFreq = inst._cache_.minFreq;
  const path = [];
  path[len] = [0, len];
  for (let s = len - 1; s >= 0; s--) {
    path[s] = [(minFreq + path[s + 1][0]), s + 1];
    for (let i = s + 1; i <= len; i++) {
      const word = sentence.slice(s, i);
      let score = pfdict[word];
      if (score) {
        score += (path[i][0]);
        if (path[s][0] < score) path[s] = [score, i];
      } else if (score === undefined) break;
    }
  }

  const words = [];
  let s = 0;
  let buf = '';
  while (s < len) {
    const e = path[s][1];
    const word = sentence.slice(s, e);
    if (word.match(/^[a-zA-Z0-9]$/)) {
      buf += word;
    } else {
      if (buf.length) {
        words.push(buf);
        buf = '';
      }
      words.push(word);
    }
    s = e;
  }
  if (buf.length) words.push(buf);
  return words;
}

async function _setup () {
  if (this._loadCache()) return;
  await this._buildPfdictFromSource();
}

function _setupSync () {
  if (this._loadCache()) return;
  this._buildPfdictFromSourceSync();
}

function _loadCache () {
  try {
    const data = fs.readFileSync(this.options.cacheFile);
    this._cache_ = JSON.parse(data);
    return true;
  } catch (e) {
    if (e.code !== 'ENOENT') {
      console.error('open ' + this.options.cacheFile + '\n', e);
      throw e;
    }
  }
  return false;
}

async function _buildPfdictFromSource () {
  this.dictionary = [];
  for (const file of this.options.dictionaryFiles) {
    const dict = await loadDict(file);
    for (let i = 0; i < dict.length; i++) this.dictionary.push(dict[i]);
  }

  if (this.options.dictionaryEntries) {
    for (let i = 0; i < this.options.dictionaryEntries.length; i++) {
      this.dictionary.push(this.options.dictionaryEntries[i]);
    }
  }

  this._buildPfdictFromDictionary(this.dictionary);
}

function _buildPfdictFromSourceSync () {
  this.dictionary = [];
  for (const file of this.options.dictionaryFiles) {
    const dict = loadDictSync(file);
    for (let i = 0; i < dict.length; i++) this.dictionary.push(dict[i]);
  }

  if (this.options.dictionaryEntries) {
    for (let i = 0; i < this.options.dictionaryEntries.length; i++) {
      this.dictionary.push(this.options.dictionaryEntries[i]);
    }
  }

  this._buildPfdictFromDictionary(this.dictionary);
}

// pfdict is an object used as a simple key/value store.
// The keys/properties are words and word prefixes
// The values are a function of the frequency of the word
// Word prefixes that are not themselves words have a frequency of zero.
// Words have a non-zero frequency.
function _buildPfdictFromDictionary (dict) {
  const pfdict = {};
  let minFreq = Infinity;
  let word;
  let freq;
  for (let i = 0; i < dict.length; i++) {
    [word, freq] = dict[i];
    pfdict[word] = freq = Math.log(freq);
    if (freq < minFreq) minFreq = freq;
    for (let j = 1; j < word.length; j++) {
      pfdict[word.slice(0, j)] ||= 0;
    }
  }

  // Update the caches
  this._cache_.pfdict = pfdict;
  this._cache_.minFreq = minFreq;
  fs.writeFileSync(this.options.cacheFile, JSON.stringify(this._cache_));
}

const api = {
  useDict,
  getDict,
  getDictFiles,
  cut,
  init,
  initSync,
  _cut,
  __cutAll,
  __cutDagNoHmm,
  _setup,
  _setupSync,
  _loadCache,
  _buildPfdictFromSource,
  _buildPfdictFromSourceSync,
  _buildPfdictFromDictionary,
};

function initInstance (options) {
  return {
    ...api,
    _cache_: {},
    dictionary: [],
    options: {
      cacheFile: path.join(os.homedir(), 'jieba-cache.json'),
      dictionaryFiles: [
        path.join(import.meta.dirname, '..', 'dict', 'dict.txt.big'),
      ],
      ...options,
    },
  };
}

export async function jiebaFactory (options = {}) {
  const instance = initInstance(options);
  await instance.init();
  return instance;
}

export function jiebaFactorySync (options = {}) {
  const instance = initInstance(options);
  instance.initSync();
  return instance;
}

export default jiebaFactory;
