// @ig3/node-jieba-js/src/syncLineReader.mjs
import fs from 'node:fs';

function getLine () {
  if (!this.fd) return null;
  let lineBuffer;
  while (1) {
    if (this.charBufferIndex >= this.charBufferLen) {
      this.charBufferLen =
        fs.readSync(this.fd, this.charBuffer, 0, this.chunkSize);
      this.charBufferIndex = 0;
      if (this.charBufferLen < this.chunkSize) {
        this.charBuffer.subarray(this.charBufferLen, this.chunkSize).fill(0);
      }
      if (this.charBufferLen === 0) break;
      if (this.trailingCR && this.charBuffer[0] === 10) { this.charBufferIndex = 1; }
      this.trailingCR = false;
    }
    const i = this.charBuffer.indexOf(10, this.charBufferIndex);
    const j = this.charBuffer.indexOf(13, this.charBufferIndex);
    if (i >= 0 || j >= 0) {
      const lineEnd = (i >= 0 && j >= 0) ? Math.min(i, j) : Math.max(i, j);
      if (lineBuffer) {
        lineBuffer = Buffer.concat([
          lineBuffer,
          this.charBuffer.subarray(this.charBufferIndex, lineEnd)
        ]);
      } else {
        lineBuffer = this.charBuffer.subarray(this.charBufferIndex, lineEnd);
      }
      this.charBufferIndex = lineEnd + 1;
      if (this.charBuffer[lineEnd] === 13) {
        if (this.charBufferIndex >= this.charBufferLen) {
          this.trailingCR = true;
        } else {
          if (this.charBuffer[this.charBufferIndex] === 10) {
            this.charBufferIndex++;
          }
          break;
        }
      } else {
        break;
      }
    } else {
      // No line termination to the end of the current charBuffer
      if (lineBuffer) {
        lineBuffer = Buffer.concat([
          lineBuffer,
          this.charBuffer.subarray(this.charBufferIndex, this.charBufferLen)
        ]);
      } else {
        lineBuffer = Buffer.from(
          this.charBuffer.subarray(this.charBufferIndex, this.charBufferLen)
        );
      }
      this.charBufferIndex = this.charBufferLen;
    }
  }
  return lineBuffer ? lineBuffer.toString(this.encoding) : null;
}

export default function syncLineReaderFactory (file) {
  const fd = fs.openSync(file);
  const chunkSize = 32768;
  return {
    fd: fd,
    getLine: getLine,
    encoding: 'utf8',
    charBuffer: Buffer.alloc(chunkSize),
    charBufferIndex: 0,
    charBufferLen: 0,
    chunkSize: chunkSize,
    trailingCR: false,
  };
}
